crates/blockchain/blockchain.rs

```diff
@@ -471,190 +486,195 @@ impl Blockchain {
+                            Some(rlp) => {
+                                let state = AccountState::decode(&rlp)?;
+                                state_trie.insert(pathrlp, rlp)?;
+                                state
+                            }
+                            None => AccountState::default(),
+                        };
+                        vacant_entry.insert(account_state);
+                    }
+                },
+                MerklizationRequest::Delete(prefix) => {
+                    tree.insert(prefix, Trie::new_temp());
                 }
-                Entry::Vacant(vacant_entry) => {
-                    let account_state = match state_trie.get(&hashed_address)? {
-                        Some(encoded_state) => {
-                            trace!(
-                                "Found account state in trie for {}",
-                                hex::encode(&hashed_address)
-                            );
-                            AccountState::decode(&encoded_state)?
-                        }
-                        None => {
-                            trace!(
-                                "Created account state in trie for {}",
-                                hex::encode(&hashed_address)
-                            );
-                            AccountState::default()
+                MerklizationRequest::MerklizeStorage { prefix, key, value } => {
+                    let trie = match tree.entry(prefix) {
+                        Entry::Occupied(occupied_entry) => occupied_entry.into_mut(),
+                        Entry::Vacant(vacant_entry) => {
+                            vacant_entry.insert(self.load_trie(parent_header, Some(prefix))?)
                         }
                     };
-                    vacant_entry.insert(account_state)
-                }
-            };
-            if update.removed_storage {
-                account_state.storage_root = *EMPTY_TRIE_HASH;
-                storage_updates_map.remove(&hashed_address_h256);
-            }
-            if let Some(info) = &update.info {
-                trace!(
-                    nonce = info.nonce,
-                    balance = hex::encode(info.balance.to_big_endian()),
-                    code_hash = hex::encode(info.code_hash),
-                    "With info"
-                );
-                account_state.nonce = info.nonce;
-                account_state.balance = info.balance;
-                account_state.code_hash = info.code_hash;
-                // Store updated code in DB
-                if let Some(code) = &update.code {
-                    trace!("Updated code");
-                    code_updates.insert(info.code_hash, code.clone());
+                    if value.is_zero() {
+                        trie.remove(&key.as_bytes().to_vec())?;
+                    } else {
+                        trie.insert(key.as_bytes().to_vec(), value.encode_to_vec())?;
```

Comment:
This can be replaced by `encode`, to reuse a single buffer across all calls (calling `clear` after each one).